core/trino-main/src/test/java/io/trino/execution/TestNodeScheduler.java

```diff
@@ -800,12 +801,67 @@ public void testEmptyAssignmentWithFullNodes()
                     .build());
         }
         unassignedSplits = Sets.difference(unassignedSplits, new HashSet<>(assignments2.values()));
-        assertEquals(unassignedSplits.size(), 10);
+        assertEquals(unassignedSplits.size(), 20); // 30 (unassignedSplits) - (10 (maxPendingSplitsPerTask) - 5(queued)) * 2 (nodes))
 
         Multimap<InternalNode, Split> assignments3 = nodeSelector.computeAssignments(unassignedSplits, ImmutableList.copyOf(taskMap.values())).getAssignments();
         assertTrue(assignments3.isEmpty());
     }
 
+    @Test
+    public void testQueueSizeAdjustment()
+    {
+        InternalNode node1 = new InternalNode("node1", URI.create("http://10.0.0.1:11"), NodeVersion.UNKNOWN, false);
+        nodeManager.addNodes(node1);
+        InternalNode node2 = new InternalNode("node2", URI.create("http://10.0.0.1:12"), NodeVersion.UNKNOWN, false);
+        nodeManager.addNodes(node2);
+
+        Set<Split> splits = new LinkedHashSet<>();
+
+        for (int i = 0; i < 20 * 9; i++) {
+            splits.add(new Split(TEST_CATALOG_HANDLE, new TestSplitLocal()));
+        }
+
+        Multimap<InternalNode, Split> assignments1 = nodeSelector.computeAssignments(splits, ImmutableList.copyOf(taskMap.values())).getAssignments();
+        assertEquals(assignments1.size(), 40);
+        MockRemoteTaskFactory remoteTaskFactory = new MockRemoteTaskFactory(remoteTaskExecutor, remoteTaskScheduledExecutor);
+        int task = 0;
+        for (InternalNode node : assignments1.keySet()) {
+            TaskId taskId = new TaskId(new StageId("test", 1), task, 0);
+            task++;
+            MockRemoteTaskFactory.MockRemoteTask remoteTask = remoteTaskFactory.createTableScanTask(taskId, node, ImmutableList.copyOf(assignments1.get(node)), nodeTaskMap.createPartitionedSplitCountTracker(node, taskId));
+            remoteTask.startSplits(assignments1.get(node).size());
+            nodeTaskMap.addTask(node, remoteTask);
+            taskMap.put(node, remoteTask);
+        }
+        Set<Split> unassignedSplits = Sets.difference(splits, new HashSet<>(assignments1.values()));
+        assertEquals(unassignedSplits.size(), 140);
+
+        Multimap<InternalNode, Split> assignments2 = nodeSelector.computeAssignments(unassignedSplits, ImmutableList.copyOf(taskMap.values())).getAssignments();
+        for (InternalNode node : assignments2.keySet()) {
```

Comment:
IMo it would be good to test/assert per node how many splits it got and test cases like doubling the number of splits for fast node, reaching the maximum of max splits per node, not increasing the max splits if node is not fast, decreasing the max splits if node is slow etc